### DISCUSSION TITLE
LDM Notes for March 30th, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5973] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-30.md

## Agenda

1. Definite assignment in struct constructors calling `: this()`
2. `file private` accessibility

**Comment by Joe4evr** [comment]
Here's a case I can think of for `file` being a modifier to accessibility in general (option 3):
```cs
public partial class MyCollection<T> : IEnumerable<T>
{
    public Enumerator GetEnumerator() => new(this);

    public struct Enumerator : IEnumerator<T>
    {
        // other files in this project shouldn't access this ctor,
        // but it needs to be internal at least for the outer type to use
        file internal Enumerator(MyCollection<T> collection) //...... 
    }
}
```

**Reply by RikkiGibson** [reply]
I think that the suggested design for option 2 (single new accessibility level) was supposed to be the most permissive one, which would make it equivalent to 'file internal', to my understanding.

**Reply by 333fred** [reply]
That was my understanding as well.

**Comment by alrz** [comment]
> 1. https://github.com/dotnet/csharplang/issues/5969 as is.
> 2. file or fileonly as a new accessibility. 
> 3. file is allowed to be combined with any existing accessibility.

Ability to combine file accessibility with others could encourage a new level of "abstraction": per file, which to be honest I don't really see as a good thing. Is this really expected to be used outside SGs? If not I'd suggest `hidden` modifier as an indication that it's not intended to be used in user code.

**Reply by CyrusNajmabadi** [reply]
> Is this really expected to be used outside SGs

Yes. A core use case for me is normal user code.

**Comment by TahirAhmadov** [comment]
For structs, I will re-iterate - please, please try to make structs work like classes - taking into account breaking existing code, like it was noted, using fix suggestions, etc. I understand why C# 1 went with parameterless struct ctors being always available, but it felt weird back then, and today, with `default(T)` being available, it's really strange. 
I'm perfectly happy with option 6; do option 4 now, but have a plan to slowly transition to option 6. If it's a struct like `Point`, which has no explicit parameterless ctor, and IDE encounters `new Point()`, it should change it to `default(Point)`. That suggestion should be rolled out as soon as possible.
Then, we can make that a warning in a future version, like C#11.

**Reply by RikkiGibson** [reply]
I think some users want the behavior of `new Point()` quietly changing to use the parameterless constructor if one is declared. I think it would be reasonable for the IDE to introduce a code style option here (prefer `default` over `new()` with the default parameterless constructor), although IDE team will have to decide whether that's the right thing to do.

**Reply by TahirAhmadov** [reply]
@RikkiGibson sure, but doesn't the site of the `new Point()` know if an explicit parameter-less ctor exists? Can that analyzer only suggest changing to `default` if no explicit ctor exists, meaning it's a style change without any change in behavior?

**Reply by TahirAhmadov** [reply]
https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBsD2BjK7VhwB22MAlpkarDAmcHDCgDQwhnrMAmIA1AD4ABAEwBGALAAoAPTTktOKQAEAQSmyA3lKWylewQGZVACgCU66Xr1aZlqwF8LjyfISKYSgEJSbVpYa9jMiIPKFMlDWdnEX9RAHYfbStBUQA2fwAWJQBZKGCzJOsk3T8lADcoBCUoJQBeJSIIAHcTUwBuQtKKquA6hubA9s6ooA===
Looks like, currently no ctor is emitted and the C# 1-style "parameterless ctor" is essentially just `default`. Obviously when explicit parameter-less ctors are allowed, then an actual ctor will have to be emitted, and the call site (instantiation site) will see that, since it's public API area. Right?

**Reply by TahirAhmadov** [reply]
Another thought - imagine this:
```cs
// hypothetical struct - no explicit parameter-less ctor
public struct Point { public Point(int x, int y) { ... } public int X, Y; }
// currently
var point = new Point() { X = 1, Y = 2, }; // works; this is basically:
var point = default(Point);
point.X = 1; point.Y = 2;
// if we start pushing users away from "new Point()"
var point = new Point() { X = 1, Y = 2, }; // this will annoy users with suggestions and even warnings in the future
// can we introduce a syntax like so?
var point = default(Point) { X = 1, Y = 2, };
```

**Reply by HaloFour** [reply]
I could see an analyzer or code style rule that prefers `default(T)` over `new T()` for structs that do not have parameterless constructors.

**Reply by RikkiGibson** [reply]
> Can that analyzer only suggest changing to `default` if no explicit ctor exists, meaning it's a style change without any change in behavior?

Yes, that's how it would work. The analyzer would probably be "info" severity by default, though, and people who want to prevent it throughout their project can set it to a higher severity.

**Reply by TahirAhmadov** [reply]
> Yes, that's how it would work. The analyzer would probably be "info" severity by default, though, and people who want to prevent it throughout their project can set it to a higher severity.

In that case I'm not sure I understand you previous post:

> I think some users want the behavior of `new Point()` quietly changing to use the parameterless constructor if one is declared. I think it would be reasonable for the IDE to introduce a code style option here (prefer `default` over `new()` with the default parameterless constructor), although IDE team will have to decide whether that's the right thing to do.

If an explicit parameter-less ctor exists, this suggestion would not apply - which makes sense obviously. If it's the `default` ctor, why would we want to keep it "legal" (or "style-acceptable") going forward? Don't we want to make it an info now, and a warning later, to get rid of that syntax altogether?
PS. Am I wrong to think that eventually (however many versions in the future), it should be an error (to match `class` behavior)?

**Reply by jnm2** [reply]
@TahirAhmadov I want to write `new S()` to mean: do all initialization you need so that I have a usable `S` instance. It's up to the struct author to determine whether this requires a parameterless constructor or not. I never want to couple my code to that implementation detail of the struct.

(I do want to get a warning when writing `new S()` would cause `default(S)` behavior even if there *was* a parameterless constructor, e.g. as an optional parameter's default value.)

**Reply by RikkiGibson** [reply]
I'd push back a little bit on calling a public constructor an implementation detail, but the described use case seems reasonable to me :)

> ```cs
> // can we introduce a syntax like so?
> var point = default(Point) { X = 1, Y = 2, };
> ```

I think LDM has considered something like this, but it just comes off strange to my eye. "I want the default value, except with these non-default values in it."

**Reply by RikkiGibson** [reply]
Opened dotnet/roslyn#60592 to track adding a code style option.

**Reply by TahirAhmadov** [reply]
> @TahirAhmadov I want to write `new S()` to mean: do all initialization you need so that I have a usable `S` instance. It's up to the struct author to determine whether this requires a parameterless constructor or not. I never want to couple my code to that implementation detail of the struct.

I would argue that it's exactly the opposite of an implementation detail. Whether a struct has a parameter-less constructor is part of its public API. If you have a generic scenario, you can use `where T:struct,new()` constraint; if the type is known, then the using site of the struct knows whether it has a parameter-less ctor or not. For advanced scenario of "call explicit parameter-less ctor if available, otherwise `default`", I envision writing some reflection-based code (IL emit or expressions compiled). This should not be a language feature.

> (I do want to get a warning when writing `new S()` would cause `default(S)` behavior even if there _was_ a parameterless constructor, e.g. as an optional parameter's default value.)

So:
```cs
struct A { public A(int x = 0) { this.X = x; } }
var a = new A(); // warning here? why? how?
struct B { public B(int x = 1) { this.X = x; } }
var b = new B(); // i guess no warning?
```
That's extremely confusing if you ask me.

**Reply by TahirAhmadov** [reply]
> I think LDM has considered something like this, but it just comes off strange to my eye. "I want the default value, except with these non-default values in it."

I guess, but it's merely a case of something looking unusual - these things are subjective and part of the reason it looks weird is it's new.
It's important to remember that this is an edge case. If you have a struct with a ctor with arguments, it's very likely meant to be used for initialization. On the other hand, if you are initializing fields/properties using an init block, it's very likely because no ctors are defined (which gives the struct an implicit public no-arg ctor, like classes) and the struct is meant to be initialized using init blocks. In the real world, I don't see this being used often, if at all. And that I think jibes well with the weirdness of the syntax: you want to do something weird, you have to use a weird syntax to accomplish it.
If we look at it on the classes side, it's basically the same as doing `FormatterServices.GetUninititializedObject`, then setting members afterwards. I think we can all agree that's not a pattern that's either common or desirable.

**Reply by TahirAhmadov** [reply]
If you have:
```cs
struct S { public int X, Y; }
var s = new S(); // this should be no problem, right?
var s2 = new S { X = 1, Y = 2, }; // and this works perfectly, as well (no need for the "weird" syntax)
```

**Reply by TonyValenti** [reply]
If it used WITH in it, I'd like it:

Default(T) with { x = 1 }

> I'd push back a little bit on calling a public constructor an implementation detail, but the described use case seems reasonable to me :)
> 
> 
> 
> > ```cs
> 
> > // can we introduce a syntax like so?
> 
> > var point = default(Point) { X = 1, Y = 2, };
> 
> > ```
> 
> 
> 
> I think LDM has considered something like this, but it just comes off strange to my eye. "I want the default value, except with these non-default values in it."

**Reply by TahirAhmadov** [reply]
> If it used WITH in it, I'd like it:
> 
> Default(T) with { x = 1 }

That would be perfect, actually - I think `with` for non-record types is being looked into, but I may be wrong.

**Reply by jnm2** [reply]
@TahirAhmadov "Implementation detail" was hyperbole. Regardless, that's the behavior I want to fall out when I say `new S()`.

**Reply by TahirAhmadov** [reply]
@jnm2 can you describe a real world use case for this?

**Reply by jnm2** [reply]
`default(Point) with { X = 1, Y = 2 }` has worked since C# 10.

The use case is to avoid bugs and code churn when a struct gains a parameterless constructor that should be called to fully initialize the struct. This hasn't really happened yet because parameterless constructors only just became available to declare in C#. There are some structs I'm planning to refactor though which used field value offsetting tricks so that `new S()` had appropriate default values for its properties.

**Reply by TahirAhmadov** [reply]
> The use case is to avoid bugs and code churn when a struct gains a parameterless constructor that should be called to fully initialize the struct.

Just so that I understand correctly, you had structs (obviously w/o no-arg ctors) and doing `new S()` was effectively `default(S)`. Now you want to modify those structs to define no-arg ctors to do some initialization. Right?
If so, I think there is a mistake being made there. The "using site" - where you have your `new S()` - used to get `default(S)` back. Now, `new S()` suddenly returns something other than `default(S)`. This means you have to modify that site to expect these changes. 
Now that we established that you're modifying (or at least reviewing) those sites, it means you can make a decision for each such line; if you need `default(S)`, you can change it (probably using IDE auto-fix); if the initialization makes sense to be reusable in struct ctor, you add that ctor and `new S()` is no longer changed to `default(S)`.
Just "assuming" that `new S()` will do what you need is wrong - exactly because you don't know what it's doing. I suspect there will be bugs caused by that assumption.

**Reply by TahirAhmadov** [reply]
Even more importantly, I think we have to decide once and for all - what is the end goal here. I'm afraid making it possible for `new S()` to be legal at all times and to mean different things depending on how a struct is defined (especially in 3rd party code) is adding a big amount of technical debt to the language. It would make it more difficult in the future to achieve a simple and consistent behavior with structs.

**Reply by HaloFour** [reply]
This behavior where `new S()` might call a constructor or might result in `default(S)` has existed since C# 2.0 in 2005.  Before then `new S()` was the only way to get a default value for a struct.  I can understand a preference for wanting to use `default(S)` instead of `new S()`, and that sounds like a good case for a code style rule or analyzer, but I don't see the compiler making an about-face about what `new S()` means at this point.

**Reply by TahirAhmadov** [reply]
> This behavior where `new S()` might call a constructor or might result in `default(S)` has existed since C# 2.0 in 2005.

How so? How can `new S()` call a ctor in a world without optional parameters and no-param ctors?

**Reply by HaloFour** [reply]
@TahirAhmadov 

> How so? How can `new S()` call a ctor in a world without optional parameters and no-param ctors?

.NET has always supported parameterless ctors on structs.  C# didn't support declaring them, but it has always supported consuming them.

**Reply by TahirAhmadov** [reply]
> .NET has always supported

Are there any .NET languages which make use of this feature? Are there any structs anywhere in BCL which have parameter-less ctors?
Those questions aside, if a parameter-less ctor does exist (however it was emitted), that's fine - keep using it. My point still stands - if there is no parameter-less ctor, doing `new S()` should be made illegal. (Slowly, taking into account breaking existing code, etc.)

**Reply by jnm2** [reply]
> Now, `new S()` suddenly returns something other than `default(S)`. This means you have to modify that site to expect these changes.

@TahirAhmadov This is actually not true and it's exactly why I was saying I do use `new S()` in this case. I do not have to modify the usage site; all it takes is a recompile. Of course you can't change `S` without getting a binary behavioral break, but that's not a concern in the cases I'm thinking of.

**Reply by TahirAhmadov** [reply]
> > Now, `new S()` suddenly returns something other than `default(S)`. This means you have to modify that site to expect these changes.
> 
> @TahirAhmadov This is actually not true and it's exactly why I was saying I do use `new S()` in this case. I do not have to modify the usage site; all it takes is a recompile. Of course you can't change `S` without getting a binary behavioral break, but that's not a concern in the cases I'm thinking of.

You don't **have** to modify the callsite from a compilation perspective, yes. But you definitely **want** to modify it (or at least review it). If the same line of code has a different outcome after a recompile, that change can have repercussions and _should not_ be ignored.

**Reply by jnm2** [reply]
I believe I can be sure I won't want to modify or review it so long as I use `new S()` ahead of time. Cases aren't coming to mind where this could be a pitfall.

**Reply by TahirAhmadov** [reply]
That's fine - you mentally analyzed your particular use case and it makes that `new S()` starts doing something different than before, and the usage site can continue working as expected. However, that's your use case only. As a general rule, it's a very dangerous way to make changes.

**Reply by RikkiGibson** [reply]
FWIW, if I introduced such a constructor, I might be inclined to "find all references" and at least skim through them.
